pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -597,16 +622,49 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 			}
 		}
 	}
-	before := pr.Status.GetCondition(apis.ConditionSucceeded)
-	after := resources.GetPipelineConditionStatus(pr, pipelineState, c.Logger, d)
-	pr.Status.SetCondition(after)
-	reconciler.EmitEvent(c.Recorder, before, after, pr)
-
+	pipelineTasksStatusBefore := pr.Status.GetCondition(apis.ConditionSucceeded)
+	pipelineTasksStatusAfter := resources.GetPipelineConditionStatus(pr, pipelineTasks, c.Logger, d)
+	// begin executing final tasks when all non-final tasks are completed successfully or any one of them have failed
+	if pipelineTasksStatusAfter.IsTrue() || (pipelineTasksStatusAfter.IsFalse() && pipelineTasksStatusAfter.Reason == resources.ReasonFailed) {
+		if err := c.executeFinalTasks(pipelineState, finalTasks, pr, pipelineSpec, pipelineTasksStatusAfter, as.StorageBasePath(pr)); err != nil {
```

Comment:
do we want to treat finally task execution as completely different or is it possible that the dag logic could be updated to understand finally tasks?

Reply:
yup its treated completely different for now, I wouldnt mix it up with dag logic, the existing dag logic doesnt handle tasks which are skipped. If a condition results in failure, a task is skipped along with all the tasks from that branch. This list of skipped tasks is determined while retrieving [PipelineRunStatus](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L446) and kept limited to the status message with its count. DAG logic is lacking means of knowing if a given task was skipped on purpose or haven't scheduled for execution yet.

Would like to propose something like [PipelineRunPipelineTaskState](https://github.com/pritidesai/pipeline/blob/skip/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L295) which holds a task with its execution status and more. This proposal will simplify dag logic, the way a list of skipped task is derived and also be beneficial for CLI and dashboard.

Reply:
merged it in the latest PR #2661